mvc/models/model-binding.rst

```diff
@@ -1,8 +1,92 @@
+
+.. note:: Form values, route data, and query strings are all stored as name-value pairs.
+
+Since model binding asked the form value provider for a key named `id` and didn't find anything, it moved on to the route values looking for a key with a name of `id`. In our example, it's a match. Binding happens, and the value of the id parameter becomes 2. So far the example use primitive types. If the action method's parameter were a complex type such as the ``Movie`` type, which contains both primitive and complex types as properties, MVC's model binding will still handle it nicely. It uses reflection and recursion to traverse the properties of complex types. While doing so, it compares primitive members' names against the names of keys in data sources and recurses through the properties that are complex types. When a value provider delivers the first item with a matching name from the data source, it stops looking for more items with that name. If binding can't occur, the error is recorded in the ``ModelState.Errors`` collection, and model binding attempts to bind to the next parameter or property. You can query for model state errors by checking the ``ModelState.IsValid`` method. 
+
+.. Note:: The model state errors are contained in a ``ModelStateDictionary`` object that contains name-value pairs of items. It's rarely necessary to query this collection yourself; if there are items in this collection, then ``ModelState.IsValid`` returns ``false``. 
+
+Additionally, there are some special data types that ASP.NET MVC must consider when performing model binding:
+
+- ``IFormFile``: Uploaded files that are part of the HTTP request.
+- ``IFormCollection``: The collection of key-value pairs that represent the form data.
+- ``CancelationToken``: Used to cancel activity in asynchronous controllers.
+
+Once model binding is complete, `validation <https://docs.asp.net/projects/mvc/en/latest/models/validation.html>`_ occurs. Default Model binding works great for the vast majority of development scenarios. It is also extensible so if you have unique needs you can customize the built-in behavior.  
+
+Customize model binding behavior with attributes 
+--------------------------------------------------------
+ASP.NET MVC contains several attributes that you can use to direct its default model binding behavior to a different source. For example, you can specify whether binding is required for a field, or if it should never happen at all by using the ``[BindRequired]`` or ``[BindNever]`` attributes. Alternatively, you can override the default value, and specify the model binder's data source, which is what the ``[From*]`` attributes do, such as the ``[FromForm]`` or ``[FromQuery]``. While the topic of model binding extensibility in depth is outside of the scope of this doc, here are a few customizations that you should be aware of:
+
+- ``[BindRequired]``: This attribute forces a model state error if binding cannot occur.
```

Comment:
If binding can’t occur, the error is recorded in the ModelState. So why is this needed?